Factory function support for NestedTensor

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #146172
* #146101
* #145922
* #141842
* #141841
* #146052

Rebase of https://github.com/pytorch/pytorch/pull/117904 removing unnecessary bits now that python nested int already holds the necessary metadata.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

soulitzer committed:
Factory function support for NestedTensor

[ghstack-poisoned]

Skylion007 commented on torch/nested/_internal/nested_int.py:
```diff
@@ -165,6 +166,11 @@ def wrap_int(self, num: int) -> ConstantIntNode:
         assert type(num) is int
         return ConstantIntNode(num)
 
+    def key_set(self) -> Any:
```
Why not make ReturnType DispatchKeySet?

Skylion007 commented on torch/nested/_internal/ops.py:
```diff
@@ -2797,3 +2799,41 @@ def _nested_get_jagged_dummy(func, *args, **kwargs):
+    ],
+    "size: any, dtype: any?, layout: any?, device: any?, pin_memory: any?",
+)
+def jagged_factory(func, *args, **kwargs):
```
You can spec args and kwargs using ParamSpec if you want and have func be a callable

soulitzer committed:
Update on "Factory function support for NestedTensor"


Rebase of https://github.com/pytorch/pytorch/pull/117904 removing unnecessary bits now that python nested int already holds the necessary metadata.


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function support for NestedTensor"


Rebase of https://github.com/pytorch/pytorch/pull/117904 removing unnecessary bits now that python nested int already holds the necessary metadata.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

soulitzer committed:
Update on "Factory function support for NestedTensor"


Rebase of https://github.com/pytorch/pytorch/pull/117904 removing unnecessary bits now that python nested int already holds the necessary metadata.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

soulitzer committed:
Update on "Factory function support for NestedTensor"


Rebase of https://github.com/pytorch/pytorch/pull/117904 removing unnecessary bits now that python nested int already holds the necessary metadata.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

soulitzer committed:
Update on "Factory function support for NestedTensor"


Rebase of https://github.com/pytorch/pytorch/pull/117904 removing unnecessary bits now that python nested int already holds the necessary metadata.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]
